oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/JdbcOAuth2AuthorizedClientService.java

```diff
@@ -0,0 +1,268 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.client;
+
+import org.springframework.context.annotation.Configuration;
+import org.springframework.dao.DataRetrievalFailureException;
+import org.springframework.jdbc.core.ArgumentTypePreparedStatementSetter;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.PreparedStatementSetter;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.oauth2.client.registration.ClientRegistration;
+import org.springframework.security.oauth2.client.registration.ClientRegistrationRepository;
+import org.springframework.security.oauth2.core.OAuth2AccessToken;
+import org.springframework.security.oauth2.core.OAuth2RefreshToken;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.EnableTransactionManagement;
+import org.springframework.transaction.annotation.Transactional;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+import org.springframework.util.StringUtils;
+
+import javax.sql.DataSource;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Timestamp;
+import java.sql.Types;
+import java.time.Instant;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * A JDBC implementation of an {@link OAuth2AuthorizedClientService}
+ * that uses a {@link JdbcTemplate} for {@link OAuth2AuthorizedClient} persistence.
+ *
+ * <p>
+ * In order to enable {@link Transactional} for this {@code OAuth2AuthorizedClientService},
+ * ensure you declare {@link EnableTransactionManagement} within a {@link Configuration}.
+ *
+ * <p>
+ * <b>NOTE:</b> This {@code OAuth2AuthorizedClientService} depends on the following
+ * table definition and therefore MUST be defined in the database schema.
+ *
+ * <pre>
+ * CREATE TABLE oauth2_authorized_client (
+ *   client_registration_id varchar(100) NOT NULL,
+ *   principal_name varchar(100) NOT NULL,
+ *   access_token_type varchar(75) NOT NULL,
+ *   access_token_value varchar(7000) NOT NULL,
+ *   access_token_issued_at timestamp NOT NULL,
+ *   access_token_expires_at timestamp NOT NULL,
+ *   access_token_scopes varchar(1000) DEFAULT NULL,
+ *   refresh_token_value varchar(7000) DEFAULT NULL,
+ *   refresh_token_issued_at timestamp DEFAULT NULL,
+ *   created_at timestamp DEFAULT CURRENT_TIMESTAMP NOT NULL,
+ *   PRIMARY KEY (client_registration_id, principal_name)
+ * );
+ * </pre>
+ *
+ * @author Joe Grandja
+ * @since 5.3
+ * @see OAuth2AuthorizedClientService
+ * @see OAuth2AuthorizedClient
+ * @see JdbcTemplate
+ */
+@Repository
+public class JdbcOAuth2AuthorizedClientService implements OAuth2AuthorizedClientService {
```

Comment:
@rwinch FYI, I have considered introducing `OAuth2AuthorizedClientDao` as it seems it would be a logical design choice, however, I felt that introducing a new interface in this instance would complicate the design and seemed a bit too much. 

If we did introduce `OAuth2AuthorizedClientDao` with the initial `JdbcOAuth2AuthorizedClientDao` implementation, what would the `OAuth2AuthorizedClientService` look like? It wouldn't do much more than delegate to the Dao. What would the name be? `DelegatingOAuth2AuthorizedClientService` would be too generic of a name since we could have another type of delegate that would delegate to another type of backing implementation other than a Dao. So then the name would have to be something like `DaoDelegatingOAuth2AuthorizedClientService`, which is still not ideal since we have `Dao` and `Service` in the name.

Now let's say we did introduce `OAuth2AuthorizedClientDao`, would we have another implementation? Possibly an O/R mapping implementation like Hibernate? I'm not convinced it's needed since we now will have JDBC. Redis is definitely a likely candidate for the next implementation but would that implement `OAuth2AuthorizedClientDao`? I don't think it's the right fit. Also, with the combination of `HttpSessionOAuth2AuthorizedClientRepository` and Spring Session backed by Redis, a user could have `OAuth2AuthorizedClient` stored in Redis today.

What are your thoughts?

Reply:
I agree. There isn't a need for a Dao